qpc/insights/auth.py

```diff
@@ -0,0 +1,159 @@
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+        params = {
+            "grant_type": GRANT_TYPE,
+            "scope": INSIGHTS_SCOPE,
+            "client_id": DISCOVERY_CLIENT_ID,
+        }
+        try:
+            logger.info(_(messages.INSIGHTS_LOGIN_REQUEST), url)
+            response = requests.post(url, headers=headers, data=params)
+        except ConnectionError as err:
+            raise InsightsAuthError(_(messages.INSIGHTS_LOGIN_REQUEST_FAILED % err))
+        except BaseHTTPError as err:
+            raise InsightsAuthError(_(messages.INSIGHTS_LOGIN_REQUEST_FAILED % err))
+
+        if response.status_code == http.HTTPStatus.OK:
+            self.auth_request = response.json()
+            logger.debug(
+                _(messages.INSIGHTS_RESPONSE), insights_sso_server, self.auth_request
+            )
+        else:
+            logger.debug(
+                _(messages.INSIGHTS_RESPONSE), insights_sso_server, response.text
+            )
+            raise InsightsAuthError(
+                _(messages.INSIGHTS_LOGIN_REQUEST_FAILED % response.reason)
+            )
+
+        return self.auth_request
+
+    def wait_for_authorization(self):
+        """Wait for the user to log in and authorize the request.
+
+        :returns: user JWT token
+        """
+        if self.auth_request:
+            device_code = self.auth_request["device_code"]
+            interval = self.auth_request.get("interval") or 5  # SSO default
+            expires_in = self.auth_request.get("expires_in") or 600  # SSO default
+
+            elapsed_time = 0
+            self.auth_token = None
+            while not self.auth_token:
+                config = read_insights_config()
+                insights_sso_server = config.get(CONFIG_SSO_HOST_KEY)
+                url = f"https://{insights_sso_server}{TOKEN_ENDPOINT}"  # Always SSL
+                headers = {"Content-Type": "application/x-www-form-urlencoded"}
+                params = {
+                    "grant_type": GRANT_TYPE,
+                    "client_id": DISCOVERY_CLIENT_ID,
+                    "device_code": device_code,
+                }
+                try:
+                    logger.debug(_(messages.INSIGHTS_LOGIN_VERIFYING), url)
+                    response = requests.post(url, headers=headers, data=params)
+                except ConnectionError as err:
+                    raise InsightsAuthError(
+                        _(messages.INSIGHTS_LOGIN_VERIFICATION_FAILED % err)
+                    )
+                except BaseHTTPError as err:
+                    raise InsightsAuthError(
+                        _(messages.INSIGHTS_LOGIN_VERIFICATION_FAILED % err)
+                    )
+
+                if response.status_code == http.HTTPStatus.OK:
+                    self.token_response = response.json()
+                    self.auth_token = self.token_response["access_token"]
```

Comment:
Accessing the dict value directly would raise a `KeyError` if `"access_token"` is not present in the response, which is different from how we access the `"error"` value a few lines later using `.get()`. Is that okay and expected?